plugin/pkg/admission/nodeenv/admission.go

```diff
@@ -0,0 +1,234 @@
+func readConfig(config io.Reader) *pluginConfig {
+	defaultConfig := &pluginConfig{}
+	if config == nil || reflect.ValueOf(config).IsNil() {
+		return defaultConfig
+	}
+	d := yaml.NewYAMLOrJSONDecoder(config, 4096)
+	for {
+		if err := d.Decode(defaultConfig); err != nil {
+			if err != io.EOF {
+				continue
+			}
+		}
+		break
+	}
+	return defaultConfig
+}
+
+// Admit merges pod's and its namespace tolerations.
+func (p *podNodeEnvironment) Admit(a admission.Attributes) error {
+	resource := a.GetResource().GroupResource()
+	if resource != api.Resource("pods") {
+		return nil
+	}
+	if a.GetSubresource() != "" {
+		// only run the checks below on pods proper and not subresources
+		return nil
+	}
+
+	obj := a.GetObject()
+	pod, ok := obj.(*api.Pod)
+	if !ok {
+		glog.Errorf("expected pod but got %s", a.GetKind().Kind)
+		return nil
+	}
+
+	nsName := a.GetNamespace()
+	var namespace *api.Namespace
+
+	namespaceObj, exists, err := p.store.Get(&api.Namespace{
+		ObjectMeta: api.ObjectMeta{
+			Name:      nsName,
+			Namespace: "",
+		},
+	})
+	if err != nil {
+		return errors.NewInternalError(err)
+	}
+	if exists {
+		namespace = namespaceObj.(*api.Namespace)
+	} else {
+		namespace, err = p.defaultGetNamespace(nsName)
+		if err != nil {
+			if errors.IsNotFound(err) {
+				return err
+			}
+			return errors.NewInternalError(err)
+		}
+	}
+	nsTolerations, err := p.getNameSpaceTolerations(namespace)
+	if err != nil {
+		return err
+	}
+
+	podTolerations, err := api.GetTolerationsFromAnnotations(pod.Annotations)
+	if err != nil {
+		return err
+	}
+
+	if conflictingTolerations(nsTolerations, podTolerations) {
+		return fmt.Errorf("namespace tolerations and pod tolerations conflict")
+	}
+
+	// modified pod tolerations = namespace tolerations + current pod tolerations
+	podTolerations = mergeTolerations(nsTolerations, podTolerations)
+
+	tolerationStr, err := json.Marshal(podTolerations)
+	if err != nil {
+		return err
+	}
+
+	pod.Annotations[api.TolerationsAnnotationKey] = string(tolerationStr)
+	return nil
+
+}
+
+func NewPodNodeEnvironment(client clientset.Interface, clusterDefaultTolerations []api.Toleration) *podNodeEnvironment {
+	// TODO: make it a shared cache to use between admission plugins
+	store := cache.NewStore(cache.MetaNamespaceKeyFunc)
+	reflector := cache.NewReflector(
+		&cache.ListWatch{
+			ListFunc: func(options api.ListOptions) (runtime.Object, error) {
+				return client.Core().Namespaces().List(options)
+			},
+			WatchFunc: func(options api.ListOptions) (watch.Interface, error) {
+				return client.Core().Namespaces().Watch(options)
+			},
+		},
+		&api.Namespace{},
+		store,
+		0,
+	)
+	return &podNodeEnvironment{
+		Handler: admission.NewHandler(admission.Create),
+		client:  client,
+		store:   store,
+		clusterDefaultTolerations: clusterDefaultTolerations,
+		reflector:                 reflector,
+	}
+}
+
+func (p *podNodeEnvironment) defaultGetNamespace(name string) (*api.Namespace, error) {
```

Comment:
feels odd to call this `defaultGetNamespace`, but then pass a name. Why not call `p.client.Core().Namespaces().Get(name)` directly? Doesn't seem to have added value.